### PR TITLE
feat(experimental): add new component `Badge`

### PR DESCRIPTION
**Pull request by vladimirpotekhin**
## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Build or CI related changes
- [ ] Tests related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes #

### CODE REVIEW

**Review by splincode — APPROVED** [review]

File: projects/demo/src/modules/experimental/badge/badge.template.html [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                Appearance
+            </ng-template>
+            <ng-template
+                documentationPropertyMode="input"
+                documentationPropertyName="value"
+                documentationPropertyType="string | number"
+                [documentationPropertyValues]="valueVariants"
+                [(documentationPropertyValue)]="value"
+            >
+                Value. If value is a number and is more than 99, it will be replaced with &laquo;99+&raquo;
```
Comment (nsbarsukov) [inline projects/demo/src/modules/experimental/badge/badge.template.html:126]: Can we use `value: number` with icon?

Is it expected behaviour (no horizontal padding around) ?
Screenshot 2023-09-01 at 11 04 50

File: projects/demo/src/modules/experimental/badge/examples/1/index.html [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+<tui-badge
+    appearance="accent"
+    value="Accent"
+    class="tui-space_horizontal-2 tui-space_vertical-2"
+></tui-badge>
+
+<tui-badge
+    appearance="dark-text"
+    value="Dark text"
+    class="tui-space_horizontal-2 tui-space_vertical-2"
```
Comment (nsbarsukov) [inline projects/demo/src/modules/experimental/badge/examples/1/index.html:16]: too many information in this example 🤔 

Should we move all `tui-space_horizontal-2` and `tui-space_vertical-2` to `Less`-file?

Something like:

```css
tui-badge {
    margin: 0.5rem;
}
```

File: projects/demo/src/modules/experimental/badge/examples/1/index.less [hunk 3]
```diff
@@ -0,0 +1,23 @@
+.outline {
```
Comment (nsbarsukov) [inline projects/demo/src/modules/experimental/badge/examples/1/index.less:1]: Is it used anywhere ?

File: projects/demo/src/modules/experimental/badge/examples/3/index.html [hunk 4]
```diff
@@ -0,0 +1,111 @@
+<div>
```
Comment (nsbarsukov) [inline projects/demo/src/modules/experimental/badge/examples/3/index.html:1]: `min-width: 18rem` ?

Screenshot 2023-09-01 at 11 18 24

File: projects/demo/src/modules/experimental/badge/badge.template.html [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+            id="sizes"
+            heading="Sizes"
+            [content]="example2"
+        >
+            <tui-badge-example-2></tui-badge-example-2>
+        </tui-doc-example>
+
+        <tui-doc-example
+            id="variants"
+            heading="Variants (mobile platform)"
```
Comment (nsbarsukov) [inline projects/demo/src/modules/experimental/badge/badge.template.html:33]:
```suggestion
            heading="Content type (mobile platform)"
```

___
The same naming as in API page
Screenshot 2023-09-01 at 11 20 02

File: projects/demo/src/modules/experimental/badge/examples/4/index.html [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<div class="container">
+    <tui-badge
+        appearance="primary"
+        size="xl"
+        value="Very long value in badge"
+    ></tui-badge>
+</div>
+
+<div class="container tui-space_top-3">
```
Comment (nsbarsukov) [inline projects/demo/src/modules/experimental/badge/examples/4/index.html:9]: Could we drop this `div`-s and put classes directly on badges ?

File: projects/experimental/components/badge/badge.component.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+        return tuiIsPresent(this.value) ? String(this.value) : '';
+    }
+
+    @HostBinding('class._empty-value')
+    get isEmpty(): boolean {
+        return !this.value && this.value !== 0;
+    }
+
+    textOverflow({offsetWidth, scrollWidth}: HTMLElement): boolean {
```
Comment (nsbarsukov) [inline projects/experimental/components/badge/badge.component.ts:86]: private ?

Comment (vladimirpotekhin) [reply projects/experimental/components/badge/badge.component.ts:86]: no, it is needed in the template

**Review by nsbarsukov — APPROVED** [review]

File: projects/demo/src/modules/experimental/badge/badge.template.html [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            </ng-template>
+        </tui-doc-documentation>
+    </ng-template>
+
+    <ng-template pageTab>
+        <ol class="b-demo-steps">
+            <li>
+                <p>
+                    Import
+                    <code>TuiBadgeModule</code>
```
Comment (waterplea) [inline projects/demo/src/modules/experimental/badge/badge.template.html:144]: How about we start to write just "Import into a module"? That way we wouldn't have to edit this part every time we create a new page :D

File: projects/experimental/components/badge/badge.style.less [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+
+            &._hoverable:hover,
+            &._hoverable:active {
+                background-color: var(--tui-neutral-bg-night-hover);
+            }
+        }
+    }
+
+    &[data-appearance='notification'] {
+        background-color: var(--tui-background-badge-notification);
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.style.less:185]: Do we have this var in open source?

File: projects/experimental/components/badge/badge.style.less [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+            &._hoverable:hover,
+            &._hoverable:active {
+                background-color: var(--tui-info-bg-night-hover);
+            }
+        }
+    }
+
+    &[data-appearance='neutral'] {
+        --left-content-color: var(--tui-neutral-fill);
+        background-color: var(--tui-neutral-bg);
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.style.less:165]: nit: If we replace all `background-color` with `background` we'll save a few bytes :)

File: projects/experimental/components/badge/badge.style.less [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+:host {
+    .transition(background-color);
+
+    --left-content-color: currentColor;
+    position: relative;
+    display: inline-flex;
+    align-items: center;
+    justify-content: center;
+    box-sizing: border-box;
+    color: var(--tui-base-01);
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.style.less:12]: Is it base-01 or text-01-night? Should it get inversed in night mode?

File: projects/experimental/components/badge/badge.style.less [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+    white-space: nowrap;
+    overflow: hidden;
+    vertical-align: middle;
+    max-width: 100%;
+
+    &:after {
+        .transition(background);
+        .fullsize();
+        content: '';
+        box-sizing: border-box;
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.style.less:23]: Does it matter?

File: projects/experimental/components/badge/badge.style.less [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+    .transition(background-color);
+
+    --left-content-color: currentColor;
+    position: relative;
+    display: inline-flex;
+    align-items: center;
+    justify-content: center;
+    box-sizing: border-box;
+    color: var(--tui-base-01);
+    border-radius: 6.25rem;
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.style.less:13]: Why so specific?

File: projects/experimental/components/badge/badge.style.less [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+
+            &._hoverable:hover {
+                background-color: var(--tui-clear-inverse-active);
+            }
+
+            &._hoverable:active {
+                background-color: var(--tui-clear-inverse-active);
+            }
+
+            &:focus:after {
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.style.less:213]: :focus-visible. Why only in this appearance/mode?

File: projects/experimental/components/badge/badge.style.less [hunk 15]
```diff
@@ -0,0 +1,461 @@
...
+        &[data-size='s'][data-platform='web'] {
+            padding: 0 0.25rem;
+        }
+    }
+
+    &[data-size='s'] {
+        height: 1.25rem;
+        min-width: 1.25rem;
+
+        .t-left-content:before {
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.style.less:264]: Why is it only set for 2 sizes? 🤔

File: projects/experimental/components/badge/badge.style.less [hunk 16]
```diff
@@ -0,0 +1,461 @@
...
+
+            &:before {
+                margin: 0;
+            }
+        }
+    }
+}
+
+.t-text {
+    font: var(--tui-font-text-s);
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.style.less:320]: Can we move this on host? So if somebody would like to make text in all badges bold they can do it easily.

File: projects/experimental/components/badge/badge.style.less [hunk 17]
```diff
@@ -0,0 +1,461 @@
...
+        content: '';
+        display: block;
+        background: currentColor;
+        margin: 0 0.125rem 0 0.25rem;
+        width: 0.25rem;
+        height: 0.25rem;
+        border-radius: 100%;
+    }
+
+    :host([data-size='l']) &:before {
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.style.less:395]: nit: I believe we use `:host[...]`, not `:host([...])` everywhere

File: projects/experimental/components/badge/badge.template.html [hunk 18]
```diff
@@ -0,0 +1,20 @@
+<div class="t-icon">
+    <ng-content></ng-content>
+</div>
+<div class="t-image">
+    <ng-content select="img"></ng-content>
+</div>
+<tui-fade
+    *ngIf="!isEmpty"
+    class="t-fade"
+    [size]="textOverflow(text) ? '1.5rem' : '0'"
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.template.html:10]: We do we need this? It shouldn't have any fade effect when it's not overflown out of the box.

File: projects/experimental/components/badge/badge.template.html [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+    (tuiResize)="(0)"
+>
+    <div
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.template.html:13]: Would love to minimize nesting if possible

File: projects/experimental/types/badge-appearance.ts [hunk 20]
```diff
@@ -0,0 +1,3 @@
+import {TuiStatus} from '@taiga-ui/kit';
+
+export type TuiBadgeAppearance = TuiStatus | 'accent' | 'dark-text';
```
Comment (waterplea) [inline projects/experimental/types/badge-appearance.ts:3]: Maybe we should inline these new types and just add them to `TuiStatus` in 4.0?

File: projects/experimental/components/badge/badge.component.ts [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Input()
+    @HostBinding('class._hoverable')
+    hoverable = false;
+
+    @HostBinding('attr.data-platform')
+    platform = this.tuiPlatform;
+
+    @HostBinding('@tuiExpressiveEntrance')
+    readonly entrance = this.animationOption;
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.component.ts:54]: Do we want it all the time? Maybe just for the notification appearance?

File: projects/experimental/components/badge/badge.style.less (hunk 11 above)
Comment (vladimirpotekhin) [reply projects/experimental/components/badge/badge.style.less:12]: seems like base-01 (with inversion) looks better

File: projects/experimental/components/badge/badge.style.less (hunk 15 above)
Comment (vladimirpotekhin) [reply projects/experimental/components/badge/badge.style.less]: old rule, removed

File: projects/experimental/components/badge/badge.component.ts (hunk 21 above)
Comment (vladimirpotekhin) [reply projects/experimental/components/badge/badge.component.ts:51]: discussed with [zergunteg](https://github.com/zergunteg), he said that this animation is for all badges (users can disable it)

File: projects/experimental/components/badge/badge.component.ts [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+    @HostBinding('attr.data-platform')
+    platform = this.tuiPlatform;
+
+    @HostBinding('@tuiPop')
+    readonly entrance = this.animationOption;
+
+    constructor(
+        @Inject(TUI_MODE) readonly mode$: Observable<TuiBrightness | null>,
+        @Inject(TUI_PLATFORM) readonly tuiPlatform: TuiPlatform,
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.component.ts:55]: nit: why tui prefix?

File: projects/experimental/components/badge/badge.component.ts (hunk 21 above)
Comment (waterplea) [reply projects/experimental/components/badge/badge.component.ts:51]: How do they disable it? Also this prop unnecessary, can move the binding to `host`

File: projects/experimental/components/badge/badge.component.ts [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+    @Input()
+    @HostBinding('attr.data-appearance')
+    appearance: TuiBadgeAppearance = 'default';
+
+    @Input()
+    @HostBinding('class._hoverable')
+    hoverable = false;
+
+    @HostBinding('attr.data-platform')
+    platform = this.tuiPlatform;
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.component.ts:48]: Unnecessary prop, you can move it to `host`. We can actually move all bindings to host and drop `@HostBinding`, I believe it makes bundle size a little bit smaller and uniformity within component is a good thing.

File: projects/experimental/components/badge/badge.template.html [hunk 24]
```diff
@@ -0,0 +1,16 @@
+<div class="t-image">
+    <ng-content select="img"></ng-content>
+</div>
+<div
+    class="t-icon"
+    [class._icon-only]="!text.innerHTML"
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.template.html:6]: `.trim()`. Also I hope it does not cause any change detection issues?

**Review by waterplea — COMMENTED** [review]
Can we add examples how to make it overflown with ellipsis and how to use fade?
```html
<tui-badge><span tuiFade>Long ass text....</span></tui-badge>
```
Also not sure this would work with current styles, we should make sure it does.

File: projects/experimental/components/badge/badge.template.html (hunk 24 above)
Comment (vladimirpotekhin) [reply projects/experimental/components/badge/badge.template.html]: no issues, I checked with angular profiler

File: projects/experimental/components/badge/badge.component.ts (hunk 21 above)
Comment (vladimirpotekhin) [reply projects/experimental/components/badge/badge.component.ts]: `@.disabled`  ?

Comment (vladimirpotekhin) [reply projects/experimental/components/badge/badge.component.ts]: after discussion we decided to remove default animation

File: projects/experimental/components/badge/badge.component.ts [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+    templateUrl: './badge.template.html',
+    styleUrls: ['./badge.style.less'],
+    providers: [MODE_PROVIDER],
+    animations: [tuiPop],
+    host: {
+        '[attr.data-platform]': 'platform',
+        '[attr.data-appearance]': 'appearance',
+        '[attr.data-size]': 'size',
+        '[class._hoverable]': 'hoverable',
+        '[tabIndex]': 'hoverable ? 0 : -1',
```
Comment (waterplea) [inline projects/experimental/components/badge/badge.component.ts:26]: If hoverable is mostly legacy or hint showing, let's not make it focusable? We can also remove focus styles.

**Review by waterplea — APPROVED** [review]

**Pull request merged**